### PR TITLE
[USER32] Load IMM table

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Load the IMM (Input Method Manager) function table in module `user32.dll`.

JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Strengthen `immtable.h` (located at `win32ss/user/user32/include`, that is a list of the `DEFINE_IMM_ENTRY` macro calls).
- Try to load the IMM table by using `immtable.h`.
- Implement the following functions by using the table: `IMPSetIMEW`, `IMPQueryIMEW`, `IMPGetIMEW`, `IMPSetIMEA`, `IMPQueryIMEA`, `IMPGetIMEA`, `SendIMEMessageExW`, `SendIMEMessageExA`, `WINNLSEnableIME`, and `WINNLSGetEnableStatus`.

### CODE REVIEW

File: win32ss/user/user32/include/immtable.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+DEFINE_IMM_ENTRY(BOOL, ImmFreeLayout, (DWORD dwFlags), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmActivateLayout, (HKL hKL), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmSetCandidateWindow, (HIMC hIMC, LPCANDIDATEFORM lpCandForm), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmConfigureIMEW, (HKL hKL, HWND hWnd, DWORD dwMode, LPVOID lpData), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmGetConversionStatus, (HIMC hIMC, LPDWORD lpdwConversion, LPDWORD lpdwSentence), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmSetConversionStatus, (HIMC hIMC, DWORD dwConversion, DWORD dwSentence), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmSetStatusWindowPos, (HIMC hIMC, LPPOINT lppt), 0)
+DEFINE_IMM_ENTRY(BOOL, ImmGetImeInfoEx, (PIMEINFOEX info, IMEINFOEXCLASS type, LPVOID pvKey), 0)
+DEFINE_IMM_ENTRY(/*PIMEDPI*/ LPVOID, ImmLockImeDpi, (HKL hKL), 0)
+DEFINE_IMM_ENTRY(/*VOID*/ BOOL, ImmUnlockImeDpi, (/*PIMEDPI*/ LPVOID dpi), 0)
```
Comment (HBelusca) [inline win32ss/user/user32/include/immtable.h:42]: Why a change of the return type?

**Review by HBelusca — COMMENTED** [review]
This looks OK & straightforward to me :)

File: win32ss/user/user32/include/immtable.h (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/user32/include/immtable.h:42]: We have to define the stub functions. But we can't define both cases of non-return and do-return by the macros.

Comment (katahiromz) [reply win32ss/user/user32/include/immtable.h:42]: Some structures are missing from `user32.dll`.

Comment (HBelusca) [reply win32ss/user/user32/include/immtable.h:42]: Well if ImmUnlockImeDpi() is not designed to return anything, then its "return" type must remain VOID in all cases... Could you show me the place in the code where using the actual VOID could create a problem?

File: win32ss/user/user32/misc/imm.c [hunk 2]
```diff
@@ -3,46 +3,35 @@
...
-
-Imm32ApiTable gImmApiEntries = {0};
 HINSTANCE ghImm32 = NULL;
 BOOL bImmInitializing = FALSE;
-BOOL ImmApiTableZero = TRUE;
 
+/* define stub functions */
+#undef DEFINE_IMM_ENTRY
+#define DEFINE_IMM_ENTRY(type, name, params, retval) \
+    static type WINAPI IMMSTUB_##name params { return (type)retval; }
```
Comment (katahiromz) [inline win32ss/user/user32/misc/imm.c:25]: This line defines the stubs. If void, then unable to return.

Comment (HBelusca) [reply win32ss/user/user32/misc/imm.c:25]: I see...
I will try to see if I could find a nice way to overcome the problem.

Comment (katahiromz) [reply win32ss/user/user32/misc/imm.c:25]:
```c
#define IMM_RETURN_VOID(retval) /* empty */
#define IMM_RETURN_NONVOID(retval) return retval
...
IMM_RETURN_##rettype
```

Comment (HBelusca) [reply win32ss/user/user32/misc/imm.c:25]: That could work, indeed :D 👍

Comment (katahiromz) [reply win32ss/user/user32/misc/imm.c]: OK, Done.

File: win32ss/user/user32/include/immtable.h (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/user32/include/immtable.h]: OK.

**Pull request merged**